### PR TITLE
Fixed #24773 -- Provide a json() method on a test response.

### PR DESCRIPTION
**Pull request by andymckay**

### CODE REVIEW

File: docs/topics/testing/tools.txt [hunk 1]
```diff
@@ -425,6 +425,20 @@ Specifically, a ``Response`` object has the following attributes:
             >>> response.context['name']
             'Arthur'
 
+    .. method:: json()
+
+        .. versionadded:: 1.8
```
Comment (coldmind) [inline docs/topics/testing/tools.txt]: 1.8 is feature-freezed.
This will come to 1.9

**Comment by MarkusH** [comment]
buildbot, test this please.

File: docs/topics/testing/tools.txt [hunk 2]
```diff
@@ -425,6 +425,20 @@ Specifically, a ``Response`` object has the following attributes:
...
+
+        The body of the response, parsed as json. Extra keyword arguments are
```
Comment (coldmind) [inline docs/topics/testing/tools.txt]: parsed as `JSON`

File: tests/test_client_regress/urls.py [hunk 3]
```diff
@@ -30,6 +30,7 @@
     url(r'^request_methods/$', views.request_methods_view),
     url(r'^check_unicode/$', views.return_unicode),
     url(r'^check_binary/$', views.return_undecodable_binary),
+    url(r'^json/$', views.return_json),
```
Comment (coldmind) [inline tests/test_client_regress/urls.py]: `json_response/` is more clear IMO

File: docs/topics/testing/tools.txt [hunk 4]
```diff
@@ -425,6 +425,20 @@ Specifically, a ``Response`` object has the following attributes:
...
+        passed to ``json.loads``, see the `Python documentation
+        <https://docs.python.org/3/library/json.html#json.loads>`_ for more
+        information. If the Content-Type is not "application/json"
```
Comment (coldmind) [inline docs/topics/testing/tools.txt]:
```
`"application/json"`
```

File: django/test/client.py [hunk 5]
```diff
@@ -641,6 +644,11 @@ def logout(self):
         logout(request)
         self.cookies = SimpleCookie()
 
+    def _parse_json(self, response, **extra):
+        if 'application/json' not in response.get('Content-Type'):
+            raise ValueError('Content-Type not set to application/json.')
+        return json.loads(response.content, **extra)
```
Comment (coldmind) [inline django/test/client.py]: How about python2 vs python3 version issues?

**Comment by andymckay** [comment]
Thanks for feedback @coldmind, all sounds good to me.

**Comment by andymckay** [comment]
buildbot, test this please.

**Comment by andymckay** [comment]
Oh what buildbot won't run tests for me? :(

**Comment by MarkusH** [comment]
buildbot, ok to test.

File: docs/spelling_wordlist [hunk 6]
```diff
@@ -394,6 +394,7 @@ JavaScript
 Jinja
 jQuery
 JServ
+json
```
Comment (timgraham) [inline docs/spelling_wordlist]: can remove?

File: docs/topics/testing/tools.txt [hunk 7]
```diff
@@ -425,6 +425,20 @@ Specifically, a ``Response`` object has the following attributes:
             >>> response.context['name']
             'Arthur'
 
+    .. method:: json()
```
Comment (timgraham) [inline docs/topics/testing/tools.txt]: json(**kwargs)

File: tests/test_client_regress/tests.py [hunk 8]
```diff
@@ -1268,6 +1268,17 @@ def test_empty_string_data(self):
         response = self.client.head('/body/', data='', content_type='application/json')
         self.assertEqual(response.content, b'')
 
+    def test_json(self):
+        # Regression test for #24773
```
Comment (timgraham) [inline tests/test_client_regress/tests.py]: I don't think a reference to the ticket is really needed (only when there is some additional context provided by the ticket that cannot be captured in comments or a docstring).

File: docs/topics/testing/tools.txt [hunk 9]
```diff
@@ -425,6 +425,20 @@ Specifically, a ``Response`` object has the following attributes:
...
+
+        .. versionadded:: 1.9
+
+        The body of the response, parsed as JSON. Extra keyword arguments are
+        passed to ``json.loads``, see the `Python documentation
```
Comment (timgraham) [inline docs/topics/testing/tools.txt]: Can make json.loads a link and drop "see the..."
```
:func:`json.loads`
```

File: docs/topics/testing/tools.txt [hunk 10]
```diff
@@ -425,6 +425,20 @@ Specifically, a ``Response`` object has the following attributes:
...
+        <https://docs.python.org/3/library/json.html#json.loads>`_ for more
+        information. If the Content-Type is not `"application/json"`
```
Comment (timgraham) [inline docs/topics/testing/tools.txt]: add backtickets around Content-Type
need to double backticks for application/json to have an effect on the rendered output

File: docs/topics/testing/tools.txt [hunk 11]
```diff
@@ -425,6 +425,20 @@ Specifically, a ``Response`` object has the following attributes:
...
+        then a ValueError will be raised on trying to parse the response.
```
Comment (timgraham) [inline docs/topics/testing/tools.txt]: add backticks around ValueError
on -> when

File: django/test/client.py [hunk 12]
```diff
@@ -641,6 +644,11 @@ def logout(self):
         logout(request)
         self.cookies = SimpleCookie()
 
+    def _parse_json(self, response, **extra):
+        if 'application/json' not in response.get('Content-Type'):
+            raise ValueError('Content-Type not set to application/json.')
```
Comment (timgraham) [inline django/test/client.py]: not set to->is not? (also, include the actual Content-Type in the message?)

File: tests/test_client_regress/tests.py [hunk 13]
```diff
@@ -1268,6 +1268,17 @@ def test_empty_string_data(self):
...
+        response = self.client.get('/json_response/')
+        self.assertEqual(response.json(), {'ford': 'prefect'})
+
+    def test_json_wrong_header(self):
+        # Regression test for #24773
+        response = self.client.get('/body/')
+        with self.assertRaises(ValueError):
```
Comment (timgraham) [inline tests/test_client_regress/tests.py]: prefer assertRaisesMessage to verify the message contents as well

**Comment by timgraham** [comment]
Please add a mention in the 1.9 release notes as well.

**Comment by timgraham** [comment]
merged in 4525a0c4669dccb3a4c5d865184f8bb8e4a89ec8, thanks!

**Comment by andymckay** [comment]
thanks for the help

**Pull request closed without merging**